src/java/org/apache/cassandra/net/FrameDecoder.java

```diff
@@ -263,9 +263,10 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws IOExceptio
     {
         if (msg instanceof BufferPoolAllocator.Wrapped)
         {
-            ByteBuffer buf = ((BufferPoolAllocator.Wrapped) msg).adopt();
+            BufferPoolAllocator.Wrapped wrapped = (BufferPoolAllocator.Wrapped) msg;
             // netty will probably have mis-predicted the space needed
-            allocator.putUnusedPortion(buf);
+            allocator.putUnusedPortion(wrapped);
+            ByteBuffer buf = ((BufferPoolAllocator.Wrapped) msg).adopt();
```

Comment:
Seems correct, but I am not sure about any potential performance implications, should think about it more to be sure